ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceAlgorithms.java

```diff
@@ -0,0 +1,89 @@
+package com.microsoft.azure.kusto.ingest;
+
+import com.microsoft.azure.kusto.data.instrumentation.FunctionOneException;
+import com.microsoft.azure.kusto.data.instrumentation.MonitoredActivity;
+import com.microsoft.azure.kusto.data.instrumentation.Tracer;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionClientException;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionServiceException;
+import com.microsoft.azure.kusto.ingest.resources.ResourceWithSas;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ResourceAlgorithms {
+    private static final int RETRY_COUNT = 3;
+    private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private ResourceAlgorithms() {
+    }
+
+    private static <TInner, TWrapper extends ResourceWithSas<TInner>, TOut> TOut resourceActionWithRetries(ResourceManager resourceManager,
+            List<TWrapper> resources, FunctionOneException<TOut, TWrapper, Exception> action, String actionName)
+            throws IngestionClientException {
+
+        if (resources.isEmpty()) {
+            throw new IngestionClientException("No resources found");
+        }
+
+        for (int i = 0; i < RETRY_COUNT; i++) {
+            if (i >= resources.size()) {
+                throw new IngestionClientException("Retry count exceeded");
+            }
+            TWrapper resource = resources.get(i);
+            try {
+
+                Map<String, String> attributes = new HashMap<>();
+                attributes.put("resource", resource.getEndpointWithoutSas());
+                attributes.put("account", resource.getAccountName());
+                attributes.put("type", resource.getClass().getName());
+                attributes.put("retry", String.valueOf(i));
```

Comment:
Should this not include some context about the resource posted to queue or uploaded to container?

Reply:
Added context